### PR TITLE
Optimize Generated Read Queries

### PR DESCRIPTION
**Pull request by do4gr**
- stop trying to fetch children if there are no parent records
- do not do queries like: `Select id from User where User.id in [...]`
- plumbing to parse and validate generated sql statements

Fixes https://github.com/prisma/prisma-engines/issues/969
Fixes https://github.com/prisma/prisma-engines/issues/1045

### CODE REVIEW

File: query-engine/core/src/interpreter/query_interpreters/nested_read.rs [hunk 1]
```diff
@@ -24,8 +26,18 @@ pub async fn m2m<'a, 'b>(
                 .projections(&parent_model_id)?
         }
     };
-
+    if parent_ids.is_empty() {
+        return Ok(ManyRecords::new(
+            query.selected_fields.names().map(|n| n.to_string()).collect(),
```
Comment (dpetrick) [inline query-engine/core/src/interpreter/query_interpreters/nested_read.rs:31]: Make this a function on `ManyRecords`, something like `pub fn empty(selected_fields: ...)`.

File: query-engine/core/src/interpreter/query_interpreters/nested_read.rs [hunk 2]
```diff
@@ -14,6 +14,8 @@ pub async fn m2m<'a, 'b>(
     let parent_field = &query.parent_field;
     let child_link_id = parent_field.related_field().linking_fields();
 
+    println!("m2m: {:?}", parent_result);
```
Comment (dpetrick) [inline query-engine/core/src/interpreter/query_interpreters/nested_read.rs:17]: Remove all printlns

File: query-engine/connector-test-kit/src/test/scala/queries/relations/UnnecessaryDBRequests.scala [hunk 3]
```diff
@@ -0,0 +1,258 @@
+package queries.relations
```
Comment (dpetrick) [inline query-engine/connector-test-kit/src/test/scala/queries/relations/UnnecessaryDBRequests.scala:1]: This test effectively does nothing and I suggest removing the entire file. It was probably important for your investigation, but adds no value beyond that.

**Review by dpetrick — CHANGES_REQUESTED** [review]

File: query-engine/core/src/interpreter/query_interpreters/nested_read.rs (hunk 2 above)
Comment (do4gr) [reply query-engine/core/src/interpreter/query_interpreters/nested_read.rs:17]: Of course, but I am not done with this PR yet.

File: query-engine/connector-test-kit/src/test/scala/queries/relations/UnnecessaryDBRequests.scala (hunk 3 above)
Comment (do4gr) [reply query-engine/connector-test-kit/src/test/scala/queries/relations/UnnecessaryDBRequests.scala:1]: Yep, that was the plan. This is only there to help me check the generated SQL, I can't think of an easy way to verify the generated statements from the Scala tests anyways.

File: query-engine/core/src/interpreter/query_interpreters/nested_read.rs (hunk 1 above)
Comment (do4gr) [reply query-engine/core/src/interpreter/query_interpreters/nested_read.rs:31]: Will do.

File: query-engine/connector-test-kit/src/test/scala/queries/relations/UnnecessaryDBRequests.scala (hunk 3 above)
Comment (do4gr) [reply query-engine/connector-test-kit/src/test/scala/queries/relations/UnnecessaryDBRequests.scala:1]: I talked a bit with Marcus and we decided to assert the number of selects being run to use this as a minimal regression test.

**Pull request merged**